Mathlib/Tactic/Congr!.lean

```diff
@@ -20,38 +20,67 @@ See the syntax docstring for more details.
 
 open Lean Meta Elab Tactic
 
-/--
-Asserts the given congruence theorem as fresh hypothesis, and then applies it.
-Return the `fvarId` for the new hypothesis and the new subgoals.
--/
-private def applyCongrThm? (mvarId : MVarId) (congrThmType congrThmProof : Expr) :
-    MetaM (List MVarId) := do
-  let mvarId ← mvarId.assert (← mkFreshUserName `h_congr_thm) congrThmType congrThmProof
-  let (fvarId, mvarId) ← mvarId.intro1P
-  let mvarIds ← mvarId.apply (mkFVar fvarId) { synthAssignedInstances := false }
-  mvarIds.mapM fun mvarId => mvarId.tryClear fvarId
-
-/--
-Try applying user-provided congruence lemmas. If any are applicable,
-returns a list of new goals.
--/
-def Lean.MVarId.userCongr? (mvarId : MVarId) : MetaM (Option (List MVarId)) :=
-  mvarId.withContext do
-    mvarId.checkNotAssigned `userCongr?
-    let some (lhs, _) := (← mvarId.getType').eqOrIff? | return none
-    let some name := lhs.getAppFn.cleanupAnnotations.constName? | return none
-    let congrTheorems := (← getSimpCongrTheorems).get name
-    -- Note: congruence theorems are in provided in decreasing order of priority.
-    for congrTheorem in congrTheorems do
-      let res ← observing? do
-        let cinfo ← getConstInfo congrTheorem.theoremName
-        let us ← cinfo.levelParams.mapM fun _ => mkFreshLevelMVar
-        let proof := mkConst congrTheorem.theoremName us
-        let ptype ← instantiateTypeLevelParams cinfo us
-        applyCongrThm? mvarId ptype proof
-      if let some mvars := res then
-        return mvars
-    return none
+initialize registerTraceClass `congr!
+initialize registerTraceClass `congr!.synthesize
+
+/-- The configuration for the `congr!` tactic. -/
+structure Congr!.Config where
+  /-- The transparency level to use when applying a congruence theorem.
+  By default this is `.reducible`, which prevents unfolding of most definitions. -/
+  transparency : TransparencyMode := TransparencyMode.reducible
+  /-- The transparency level to use when doing transformations before applying congruence lemmas.
+  This includes trying to prove the goal by `rfl` and using the `assumption` tactic.
+  By default this is `.reducible`, which prevents unfolding of most definitions. -/
+  preTransparency : TransparencyMode := TransparencyMode.reducible
+  /-- For passes that synthesize a congruence lemma using one side of the equality,
+  we run the pass both for the left-hand side and the right-hand side. If `preferLHS` is `true`
+  then we start with the left-hand side.
+
+  This can be used to control which side's definitions are expanded when applying the
+  congruence lemma (if `preferLHS = true` then the RHS can be expanded). -/
+  preferLHS : Bool := true
+  /-- Allow one or both sides to be a partial application.
+  When false, given an equality `f a b = g x y` this means we never consider
+  proving `f a = g x`. -/
```

Comment:
Do we still consider proving `f = g`?

Reply:
Yes, and I rewrote this to try to clarify things.